### ISSUE TITLE
pnpm update some-dep updates unrelated transitive dependencies

### ISSUE THREAD

**Issue by skovhus** [pnpm/pnpm#6657]
### pnpm version:
8.6.1

### Code to reproduce the issue:
See https://github.com/skovhus/pnpm-update-transitive-bug

`pnpm update --latest prettier` (or any other specific dependency) unexpectedly upgrades transient dependencies that are unrelated to the specified dependency. Note that `--latest` is not required for this to fail and even `pnpm update prettier@2.8.0` also updates unrelated dependencies.

```json
{
  "dependencies": {
    "@storybook/react": "7.0.13",
    "prettier": "2.8.0"
  },
  "engines": {
    "pnpm": "8.x"
  }
}
```

### Expected behavior:

Only the specified dependency (and it's transitive dependencies) are updated.

### Actual behavior:

Unrelated transitive dependencies are updated.

### Additional information:

 - `node -v` prints: tried with 18.15.0 and v16.19.1 
 - Windows, macOS, or Linux?: macOS

**Comment by skovhus** [comment]
I would like to help here, but not sure where to start on this. @zkochan

**Comment by corydeppen** [comment]
This seems similar to #5552. Unfortunately, I'm finding the `update` command to be problematic, especially when other tools like the Browserslist updater issue a command like `pnpm up caniuse-lite` that results in a ton of unrelated/unexpected changes in `pnpm-lock.yaml`. In my case, the changes in the lock file appear to be related to dependencies in the root of the monorepo where the command is being run - other workspace packages don't appear to be affected in the same way.

**Comment by rdsedmundo** [comment]
I just experienced the exact similar issue as @corydeppen, running `npx update-browserslist-db@latest` which internally calls `pnpm up caniuse-lite` caused `jest` to be upgraded which is totally non-related and it introduced a regression in our tests (due to a Jest bug).

**Comment by boliveira** [comment]
Just noticed this behaviour as well, will this be addressed?

**Comment by cdignam-segment** [comment]
The workaround I've found is to use `pnpm add my-package@latest` instead of `pnpm update my-package`

**Comment by zkochan** [comment]
Looks like this happens to the automatically installed peer dependencies. As a workaround for now, you can install those peer dependencies as dev dependencies in your project. In that case the won't be updated.

**Comment by ToshB** [comment]
I'm seeing the same issue in my monorepo.

If I run `pnpm up -r pkgA@latest` it updates that single package as expected. 
If i use `pnpm up -r --latest pkgA` (or `pnpm up -r -i --latest` and pick that specific package) some unrelated _optional_ peerDependency (in my case eslint) of a totally different package also gets updated. In my case it's eslint every time, which I guess could indicate a conflict somewhere. 

The strange thing is I am trying to constrain the eslint version via pnpm.overrides, which seems to be ignored when updating with the latest-flag. If I rerun `pnpm i` after updating nothing changes. But if I make _any_ changes to the root package.json deps/devDeps (which I guess invalidates some pnpm cache), the next time I run `pnpm i` the version of eslint is corrected to match the override. Which might indicate that the update/latest command is skipping some step?

**Comment by romanstetsyk** [comment]
This behavior occurs in the latest version as of now, `pnpm@9.10.0`.

**Comment by tho-graf** [comment]
I think i ran into this issue as well. I updated @mui/material-ui via `pnpm update -i --latest` and chose only `@mui/material-ui` in the dialog. But at the end I updated tons of deps like eslint etc.

When i ran `pnpm update @mui/material-ui@6.1.1` everthing worked fine.

**Comment by zkochan** [comment]
:ship: 9.12.1

**Comment by JasonKleban** [comment]
I think I'm seeing this still on 9.12.3.  I update only specific dependencies in the interactive chooser  `pnpm up --latest -i -r`.  Running `pnpm up --latest -i -r` again gives "All of your dependencies are already up to date" which is NOT expected or desired.  Of course, I want to stay up to date with all dependencies but I'd like to isolate certain updates to a commit or to a PR and do the rest later.

**Comment by xzhdev** [comment]
Could you please confirm whether the peer dependencies update issue has been resolved in version v10.17.0? I'm currently using v10.10.0 and still experiencing unwanted updates to unrelated dependencies, even after installing the peer dependencies as devDependencies in the project. @zkochan
